### PR TITLE
HTML-784: Don't match existing obs groups to obs groups on a form if …

### PR DESCRIPTION
**Pull request by mogoodrich**
…a concept/answer pair doesn't match

## Issue I worked on

see https://issues.openmrs.org/browse/HTML-

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by mogoodrich** [comment]
https://issues.openmrs.org/browse/HTML-784

File: api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java [hunk 1]
```diff
@@ -640,13 +638,16 @@ public Obs removeExistingObs(Concept question, Boolean answer) {
...
 		for (Map.Entry<Obs, Set<Obs>> e : existingObsInGroups.entrySet()) {
-			if (path.equals(ObsGroupComponent.getObsGroupPath(e.getKey()))) {
-				if (ret == null)
+			if (path.equals(ObsGroupComponent.getObsGroupPath(e.getKey()))
+			        && !(ObsGroupComponent.supportingRank(questionsAndAnswers, e.getValue()) < 0) // don't match an obs group if it's explicitly ranked as NOT a match for this obs group (ie has a rank < 0)
+			) {
+				if (ret == null) {
 					ret = e.getKey();
+				}
 				unmatchedContenterCount++;
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java:651]: This logic is pretty hard to understand (though it's not part of this PR... just if you feel inspired to improve the names a bit...)

**Review by brandones — APPROVED** [review]
LGTM, though a review from someone who's more familiar with this code would be valuable.

File: api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java (hunk 1 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java:651]: Agreed that it's hard to understand...

**Comment by mogoodrich** [comment]
Thanks @brandones !
